visualization/app/codeCharta/ui/export3DMapButton/export3DMapButton.component.ts

```diff
@@ -0,0 +1,38 @@
+import "./export3DMapButton.component.scss"
+import { Component, Inject } from "@angular/core"
+import { FileDownloader } from "../../util/fileDownloader"
+import { STLExporter } from "three/examples/jsm/exporters/STLExporter"
+import { ThreeSceneService } from "../codeMap/threeViewer/threeSceneService"
+import { FileNameHelper } from "../../util/fileNameHelper"
+import { isDeltaState } from "../../model/files/files.helper"
+import { Store } from "../../state/angular-redux/store"
+import { accumulatedDataSelector } from "../../state/selectors/accumulatedData/accumulatedData.selector"
+import { filesSelector } from "../../state/store/files/files.selector"
+import { FileState } from "../../model/files/files"
+import { Mesh } from "three"
+
+@Component({
+	selector: "cc-export-threed-map-button",
+	template: require("./export3DMapButton.component.html")
+})
+export class Export3DMapButtonComponent {
+	private fileName: string
+	private files: FileState[]
+	exporter = new STLExporter()
+
+	constructor(@Inject(Store) store: Store) {
+		store.select(accumulatedDataSelector).subscribe(accumulatedData => {
+			this.fileName = accumulatedData.unifiedFileMeta?.fileName
+		})
+		store.select(filesSelector).subscribe(files => {
+			this.files = files
+		})
+	}
+
+	downloadStlFile() {
```

Comment:
I would add some tests for this logic. Some of my ideas: Propably for edge cases like Empty ThreeMesh, One Selected File, Multiple Selected Files, Delta File. And I would test that the right stl file will be downloaded if the mode changes or another map was uploaded.

Reply:
I would love some help with that. I tried but didn't succeed, as I was constantly stuck.

Reply:
The download doesn't depent on the CodeCharta state, does it? As far as I understand it, it takes the Three.js model and generates a binary for the 3d printer out of it. Therefore we don't need to test different CodeCharta states like multiple or delta mode...

I would write a test that the FileDonloader gets called. The FileDownlaoder then could be tested through pure unit tests, I hope.

Reply:
Ah, I see. You are right!
@shaman-apprentice Does unifiedFileMeta mean that it is a combination of all selected files?

Reply:
yes. But I am not to sure about the naming and calculation - it is extracted by one of the services into a selector, as it is purely dependend from store state and can then easily be used by Angular components (and has better separations of concers, as the services do too much anyway)

Reply:
We added a test which verifies the template and controller binding. 

The download logic is still untested. It uses mostly external libs which are awkward to test with current design. We discussed a better approach, which is possible after Angular migration. Current implemented approach is still better than our "normal" Controller tests. According to the boy scouts rule I am OKish with it for now  (although the rule only kind of applies, as it is new code).

Reply:
Personally, I would try to spy on the STLExporter:parse() and the FileDownloader:downloadData() function. Then, I would just test that the file name is set correctly.

Reply:
I tested locally to download the STL file for 5 selected maps in multiple mode. The resulting file name is very long: `file_aggregation_of_sample1_and_sample2_and_codecharta_2021-12-16_14-52_and_junit4.source.modified_and_codecharta_edges_and_codecharta_analysis_2021-12-18_12-45.stl`. 

@BridgeAR Do you have any requirement for that?